src/DataSource.ts

```diff
@@ -91,12 +91,8 @@ export class AlertmanagerDataSource extends DataSourceApi<CustomQuery, GenericOp
     const fields = [{ name: 'Time', type: FieldType.time }];
 
     if (data.length > 0) {
-      const annotations: string[] = data
-        .map((alert: any) => Object.keys(alert.annotations))
-        .reduce((data: string[][]) => data.flat());
-      const labels: string[] = data
-        .map((alert: any) => Object.keys(alert.labels))
-        .reduce((data: string[][]) => data.flat());
+      const annotations: string[] = data.map((alert: any) => Object.keys(alert.annotations)).flat();
```

Comment:
Neither of the variants is correct.

```
// given
data=[{annotations:{foo: 1, bar: 2}}, {annotations:{bar:2, baz:3}}, {annotations:{foo:1,bar:3}}]

// before
data.map((alert) => Object.keys(alert.annotations)).reduce((data) => data.flat()) // [ "foo", "bar" ]

// after
data.map((alert) => Object.keys(alert.annotations)).flat() // [ "foo", "bar", "bar", "baz", "foo", "bar" ]

// what it really must return:
[ "foo", "bar", "baz"]
```

Reply:
How about something like that instead?

```
if (data.length > 0) {
    const attributes = new Set(['alertstatus', 'alertstatus_code']);
    for (const alert of data) {
        for (const key of Object.keys(alert.annotations)) {
            attributes.add(key);
        }
        for (const key of Object.keys(alert.labels)) {
            labels.add(key);
        }
    }

    attributes.forEach((attribute: string) => {
        fields.push({
            name: attribute,
            type: FieldType.string,
        });
    });
}
```

Reply:
@d-s-x, please take a look at line 97 where a Set is created
```
const attributes: string[] = [...new Set([...annotations, ...labels, ...alertstatus])];
```
All duplicates will be removed and the resulting array will contain the correct values

Reply:
BTW, your solution is good as well and I do not mind approving it when there will be a PR.

Reply:
> the resulting array will contain the correct values

Maybe I'm not using word "correct" correctly. The code works, yes. But it requires creation of temporary array that is larger than array of alerts. If an alert have 20 attributes on average then the attribute array will be 20 times larger. In other words this requires O(n*m) space instead of O(m) (where n is number of alerts and m is number of attributes)

Reply:
> BTW, your solution is good as well and I do not mind approving it when there will be a PR.

Well, actually I've changed order of attributes (Set retains insertion order) so my code should not be used as-is. I just wanted to illustrate the issue.

Reply:
Actually, I do not think that we should care about the order that much, as there are Grafana transformers that can do the post-ordering of the attributes and labels.
Note, that the current solution (the one that is broken and my) does not care about the order as well - it just collects them as they appear. Later they all are added into a Set that retains the order again.
So, both solutions are not about the insertion order :)